### PR TITLE
Refactor Marlin MoeRunner

### PR DESCRIPTION
**Pull request by trangdough**
## Motivation

Refactor Marlini MoE runner integration into `marlin.py` per https://github.com/sgl-project/sglang/issues/8715

## Modifications

- Implemented `MarlinRunnerInput`, `MarlinRunnerOutput`, `MarlinMoeQuantInfo`, `MarlinRunnerCore`, `pre_permute_standard_to_marlin`, and `post_permute_marlin_to_standard`

- Added `TestAWQMarlinFloat16` to `test_awq.py`

## Accuracy Tests

- Passed `sglang/test/srt/quant/test_awq.py` and `sglang/test/srt/test_gptqmodel_dynamic.py` on B200

## Benchmarking and Profiling

N/A

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).
- [x] Work with maintainers to merge your PR. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process)

### CODE REVIEW

File: python/sglang/srt/layers/quantization/gptq.py [hunk 1]
```diff
@@ -1053,47 +1055,64 @@ def create_moe_runner(
...
+            w2_qweight=layer.w2_qweight,
+            w13_scales=layer.w13_scales,
+            w2_scales=layer.w2_scales,
+            w13_g_idx=layer.w13_g_idx,
+            w2_g_idx=layer.w2_g_idx,
+            w13_g_idx_sort_indices=layer.w13_g_idx_sort_indices,
+            w2_g_idx_sort_indices=layer.w2_g_idx_sort_indices,
+            weight_bits=self.quant_config.weight_bits,
+            is_k_full=self.is_k_full,
+            workspace=layer.workspace,
```
Comment (ch-wan) [inline python/sglang/srt/layers/quantization/gptq.py]: Semantically, `workspace` should be a member of runner backend rather than its input. We can move `marlin_make_workspace` to backend init.

File: python/sglang/srt/layers/moe/moe_runner/marlin.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+            fused_marlin_moe,
+        )
+
+        x = runner_input.hidden_states
+
+        assert self.config.activation == "silu", "Only SiLU activation is supported."
+
+        orig_dtype = x.dtype
+        # removed since fused_marlin_moe.py now accepts bfloat16
+        # x = x.half()
```
Comment (ch-wan) [inline python/sglang/srt/layers/moe/moe_runner/marlin.py]: can be removed

File: python/sglang/srt/layers/quantization/gptq.py [hunk 3]
```diff
@@ -1053,47 +1055,64 @@ def create_moe_runner(
...
-            g_idx2=layer.w2_g_idx,
-            sort_indices1=layer.w13_g_idx_sort_indices,
-            sort_indices2=layer.w2_g_idx_sort_indices,
-            num_bits=self.quant_config.weight_bits,
-            is_k_full=self.is_k_full,
-        ).to(orig_dtype)
-        return StandardCombineInput(hidden_states=output)
+        return self.runner.run(dispatch_output, quant_info)
+
+        # from sglang.srt.layers.moe.fused_moe_triton.fused_marlin_moe import (
```
Comment (ch-wan) [inline python/sglang/srt/layers/quantization/gptq.py]: draft code

File: python/sglang/srt/layers/quantization/gptq.py [hunk 4]
```diff
@@ -1053,47 +1055,64 @@ def create_moe_runner(
         self, layer: torch.nn.Module, moe_runner_config: MoeRunnerConfig
     ):
         self.moe_runner_config = moe_runner_config
+        self.runner = MoeRunner(MoeRunnerBackend.MARLIN, moe_runner_config)
```
Comment (ch-wan) [inline python/sglang/srt/layers/quantization/gptq.py:1063]: add assertion: `get_moe_runner_backend().is_auto()`

File: python/sglang/srt/layers/quantization/awq.py [hunk 5]
```diff
@@ -826,43 +825,28 @@ def create_moe_runner(
         self, layer: torch.nn.Module, moe_runner_config: MoeRunnerConfig
     ):
         self.moe_runner_config = moe_runner_config
+        self.runner = MoeRunner(MoeRunnerBackend.MARLIN, moe_runner_config)
```
Comment (ch-wan) [inline python/sglang/srt/layers/quantization/awq.py:832]: add assertion: get_moe_runner_backend().is_auto()

**Review by ch-wan — COMMENTED** [review]
LGTM in general. I just left some minor comments. Thank you for your contribution.

File: python/sglang/srt/layers/moe/moe_runner/marlin.py [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+    from sglang.srt.layers.moe.fused_moe_triton.fused_marlin_moe import fused_marlin_moe
+    from sglang.srt.layers.moe.token_dispatcher.standard import StandardCombineInput
+    from sglang.srt.layers.quantization.marlin_utils import marlin_make_workspace
+
+    hidden_states = dispatch_output.hidden_states
+    topk_output = dispatch_output.topk_output
+
+    assert runner_config.activation == "silu", "Only SiLU activation is supported."
+
+    workspace = marlin_make_workspace(hidden_states.device, max_blocks_per_sm=4)
```
Comment (ch-wan) [inline python/sglang/srt/layers/moe/moe_runner/marlin.py]: `marlin_make_workspace` has a `torch.zero()` operation, which requires to launch a cuda kernel. https://github.com/sgl-project/sglang/blob/9a426fc5ef9d8b93f8a12cd6425b27b90852f9f1/python/sglang/srt/layers/quantization/marlin_utils.py#L266

We can create a global static buffer as you did in `MarlinRunnerCore.run` to avoid it.

File: python/sglang/srt/layers/moe/moe_runner/marlin.py [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+    # Optional
+    expert_map: Optional[torch.Tensor] = None
+
+
+class MarlinRunnerCore(MoeRunnerCore):
+    def __init__(self, config: MoeRunnerConfig):
+        super().__init__(config)
+        self.workspace: Optional[torch.Tensor] = None
+
+    def run(
```
Comment (ch-wan) [inline python/sglang/srt/layers/moe/moe_runner/marlin.py]: This function is a dead script, and it should not directly call `fused_marlin_moe` as it contains a permute operation. We can remove this function.

File: python/sglang/srt/layers/moe/moe_runner/marlin.py (hunk 6 above)
Comment (ch-wan) [inline python/sglang/srt/layers/moe/moe_runner/marlin.py]: `marlin_make_workspace` has a `torch.zero()` operation, which requires to launch a cuda kernel. https://github.com/sgl-project/sglang/blob/9a426fc5ef9d8b93f8a12cd6425b27b90852f9f1/python/sglang/srt/layers/quantization/marlin_utils.py#L266

We can create a global static buffer `MARLIN_MOE_WORKSPACE` as you did in `MarlinRunnerCore.run` to avoid its repeated initilization.


```suggestion
    if MARLIN_MOE_WORKSPACE is None or MARLIN_MOE_WORKSPACE.device != hidden_states.device:
        MARLIN_MOE_WORKSPACE = marlin_make_workspace(hidden_states.device, max_blocks_per_sm=4)
```

File: python/sglang/srt/layers/moe/moe_runner/marlin.py [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+    w2_qzeros: Optional[torch.Tensor] = None
+
+    # Optional
+    expert_map: Optional[torch.Tensor] = None
+
+
+class MarlinRunnerCore(MoeRunnerCore):
+    def __init__(self, config: MoeRunnerConfig):
+        super().__init__(config)
+        self.workspace: Optional[torch.Tensor] = None
```
Comment (ch-wan) [inline python/sglang/srt/layers/moe/moe_runner/marlin.py]: can be removed

**Review by ch-wan — APPROVED** [review]
LGTM
